app/src/main/java/com/lagradost/cloudstream3/utils/BiometricAuthenticator.kt

```diff
@@ -0,0 +1,177 @@
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+
+                val authFlag = DEVICE_CREDENTIAL or BIOMETRIC_WEAK or BIOMETRIC_STRONG
+                promptInfo = BiometricPrompt.PromptInfo.Builder()
+                    .setTitle(activity.getString(title))
+                    .setDescription(description)
+                    .setAllowedAuthenticators(authFlag)
+                    .build()
+            } else {
+                @Suppress("DEPRECATION")
+                promptInfo = BiometricPrompt.PromptInfo.Builder()
+                    .setTitle(activity.getString(title))
+                    .setDescription(description)
+                    .setDeviceCredentialAllowed(true)
+                    .build()
+            }
+        } else {
+            // fallback for A12+ when both fingerprint & Face unlock is absent but PIN is set
+            @Suppress("DEPRECATION")
+            promptInfo = BiometricPrompt.PromptInfo.Builder()
+                .setTitle(activity.getString(title))
+                .setDescription(description)
+                .setDeviceCredentialAllowed(true)
+                .build()
+        }
+    }
+
+    private fun isBiometricHardWareAvailable(): Boolean {
+        var result = false
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+            when (biometricManager.canAuthenticate(
+                DEVICE_CREDENTIAL or BIOMETRIC_STRONG or BIOMETRIC_WEAK
+            )) {
+                BiometricManager.BIOMETRIC_SUCCESS -> result = true
+                BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> result = false
+                BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> result = false
+                BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> result = false
+                BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED -> result = true
+                BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED -> result = true
+                BiometricManager.BIOMETRIC_STATUS_UNKNOWN -> result = false
+            }
+        } else {
+            @Suppress("DEPRECATION")
+            when (biometricManager.canAuthenticate()) {
+                BiometricManager.BIOMETRIC_SUCCESS -> result = true
+                BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE -> result = false
+                BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE -> result = false
+                BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED -> result = false
+                BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED -> result = true
+                BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED -> result = true
+                BiometricManager.BIOMETRIC_STATUS_UNKNOWN -> result = false
+            }
+        }
+
+        return result
+    }
+
+    // only needed for Android 9 and below
+    fun deviceHasPasswordPinLock(context: Context?): Boolean {
+        val keyMgr =
+            context?.getSystemService(AppCompatActivity.KEYGUARD_SERVICE) as KeyguardManager
```

Comment:
This cast throws an exception if context is null, either handle that with as? or dont allow nullable argument

Reply:
done